bridge/github/config.go

```diff
@@ -169,63 +169,112 @@ func (*Github) ValidateConfig(conf core.Configuration) error {
+	return pollGithubUntilUserAuthorizedGitbug(&values)
+}
 
+func requestUserVerificationCode(scope string) (*http.Response, error) {
+	params := url.Values{}
+	params.Set("client_id", GithubClientID)
+	params.Set("scope", scope)
 	client := &http.Client{
 		Timeout: defaultTimeout,
 	}
-
-	return client.Do(req)
+	resp, err := client.PostForm("https://github.com/login/device/code", params)
+	if err != nil {
+		return nil, err
+	}
+	if resp.StatusCode != http.StatusOK {
+		defer resp.Body.Close()
+		bb, _ := ioutil.ReadAll(resp.Body)
+		return nil, fmt.Errorf("error creating token %v: %v", resp.StatusCode, string(bb))
+	}
+	return resp, nil
 }
 
-func decodeBody(body io.ReadCloser) (string, error) {
-	data, _ := ioutil.ReadAll(body)
-
-	aux := struct {
-		Token string `json:"token"`
-	}{}
+func promptUserToGoToBrowser(code string) {
+	fmt.Println("Please visit the following URL in a browser and enter the user authentication code.")
+	fmt.Println()
+	fmt.Println("  URL: https://github.com/login/device")
+	fmt.Println("  user authentiation code: ", code)
+	fmt.Println()
+}
 
-	err := json.Unmarshal(data, &aux)
-	if err != nil {
-		return "", err
+func pollGithubUntilUserAuthorizedGitbug(values1 *url.Values) (string, error) {
+	params := url.Values{}
+	params.Set("client_id", GithubClientID)
+	params.Set("device_code", values1.Get("device_code"))
+	params.Set("grant_type", "urn:ietf:params:oauth:grant-type:device_code") // fixed by RFC 8628
+	client := &http.Client{
+		Timeout: defaultTimeout,
 	}
-
-	if aux.Token == "" {
-		return "", fmt.Errorf("no token found in response: %s", string(data))
+	// there exists a minimum interval required by the github API
+	var initialInterval time.Duration = 6 // seconds
+	var interval time.Duration = initialInterval
+	token := ""
+	for {
+		resp, err := client.PostForm("https://github.com/login/oauth/access_token", params)
+		if err != nil {
+			return "", err
+		}
+		defer resp.Body.Close()
+		if resp.StatusCode != http.StatusOK {
+			bb, _ := ioutil.ReadAll(resp.Body)
+			return "", fmt.Errorf("error creating token %v, %v", resp.StatusCode, string(bb))
+		}
+		data2, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return "", err
+		}
+		values2, err := url.ParseQuery(string(data2))
+		if err != nil {
+			return "", err
+		}
+		apiError := values2.Get("error")
+		if apiError != "" {
+			if apiError == "slow_down" {
+				interval *= 2
+			} else {
+				interval = initialInterval
+			}
+			if apiError == "authorization_pending" || apiError == "slow_down" {
+				// no-op
+			} else {
+				// apiError equals on of: "expired_token", "unsupported_grant_type",
+				// "incorrect_client_credentials", "incorrect_device_code", or "access_denied"
+				return "", fmt.Errorf("error creating token %v, %v", apiError, values2.Get("error_description"))
+			}
+			time.Sleep(interval * time.Second)
+			continue
+		}
+		token = values2.Get("access_token")
+		if token == "" {
+			panic("invalid Github API response")
```

Comment:
panic should really be only for extreme cases. Here, somehow Github return an unexpected result. An error will do fine.